src/parser/statement.js

```diff
@@ -1009,21 +1019,43 @@ export default class StatementParser extends ExpressionParser {
       return;
     }
 
+    const isPrivate = this.match(tt.hash);
+    if (isPrivate) {
+      this.expectOnePlugin(["classPrivateProperties", "classPrivateMethods"]);
+      // Skip over the hash.
+      this.next();
+    }
+
     const isSimple = this.match(tt.name);
-    const key = this.parseClassPropertyName(methodOrProp);
+    const key = isPrivate
+      ? this.parseClassPrivateName(methodOrProp)
+      : this.parseClassPropertyName(methodOrProp);
 
     this.parsePostMemberNameModifiers(methodOrProp);
 
     if (this.isClassMethod()) {
+      method.kind = "method";
+
+      if (isPrivate) {
+        // TODO: seems a bit inconsistent error throwing: in some cases we've already
+        // eaten the #, but in other cases we haven't. How to solve?
```

Comment:
You could make `expectPlugin` accept an optional `pos` parameter (like `unexpected` does) and pass `node.start` to it.